File: ios/GradeThread/Import/ImportField.swift
import Foundation

/// US-667 — the FlipDesk fields a CSV column can map to. Mirrors the web's
/// `IMPORT_FIELDS` (`src/lib/import-mapping.ts`), trimmed to what the iOS insert
/// path writes. `.skip` excludes a column. `guessField` auto-maps headers.
enum ImportField: String, CaseIterable, Identifiable, Equatable {
    case skip
    case sku
    case title
    case brand
    case style
    case size
    case color
    case material
    case conditionNotes
    case category
    case purchasePrice
    case listPrice
    case purchaseDate
    case status

    var id: String { rawValue }

    var label: String {
        switch self {
        case .skip:           return "— Skip —"
        case .sku:            return "Item # (SKU)"
        case .title:          return "Item title"
        case .brand:          return "Brand"
        case .style:          return "Style"
        case .size:           return "Size"
        case .color:          return "Color"
        case .material:       return "Material"
        case .conditionNotes: return "Notes"
        case .category:       return "Category"
        case .purchasePrice:  return "Purchase price"
        case .listPrice:      return "List price"
        case .purchaseDate:   return "Purchase date"
        case .status:         return "Status"
        }
    }

    /// Title is the only required field — a row without it can't be inserted.
    var isRequired: Bool { self == .title }

    /// Best-guess mapping from a sheet header (any casing/spacing). Mirrors the
    /// web `guessField`; unknown headers map to `.skip`.
    static func guess(from header: String) -> ImportField {
        let key = header.lowercased().filter { $0.isLetter || $0.isNumber }
        let table: [String: ImportField] = [
            "item": .sku, "itemnumber": .sku, "itemno": .sku, "sku": .sku,
            "itemtitle": .title, "title": .title, "name": .title,
            "brand": .brand,
            "style": .style, "model": .style,
            "size": .size,
            "color": .color, "colour": .color,
            "material": .material, "fabric": .material,
            "notes": .conditionNotes, "note": .conditionNotes,
            "description": .conditionNotes, "desc": .conditionNotes,
            "category": .category, "cat": .category,
            "purchaseprice": .purchasePrice, "cost": .purchasePrice, "paid": .purchasePrice,
            "listprice": .listPrice, "price": .listPrice, "targetprice": .listPrice,
            "purchasedate": .purchaseDate, "purchased": .purchaseDate, "acquired": .purchaseDate,
            "status": .status,
        ]
        return table[key] ?? .skip
    }
}

/// Pure value-parsing helpers shared by the mapper.
enum ImportValue {
    /// Parses a price-ish string ("$19.99", "1,299", "(5.00)", "1.299,00") to a
    /// Double. US-1162: a CSV can be in any locale, so rather than blindly
    /// stripping everything but ".", we treat the right-most '.'/',' as the
    /// decimal separator and the other as grouping. That parses both
    /// "1,299" -> 1299 and "1.299,00" -> 1299.00 (the old filter corrupted the
    /// latter to 1.29900).
    static func price(_ raw: String?) -> Double? {
        guard let raw else { return nil }
        // Accounting-style negatives are wholly wrapped in parens ("(5.00)").
        // A parenthetical note in a price cell ("$20 (sale)") must NOT flip the
        // sign, so require the trimmed value to start "(" and end ")".
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        let negative = trimmed.hasPrefix("(") && trimmed.hasSuffix(")")
        var s = String(raw.filter { $0.isNumber || $0 == "." || $0 == "," || $0 == "-" })
        guard !s.isEmpty else { return nil }

        let lastDot = s.lastIndex(of: ".")
        let lastComma = s.lastIndex(of: ",")
        if let dot = lastDot, let comma = lastComma {
            // Both present: the right-most is the decimal separator.
            let decimalIsDot = dot > comma
            s.removeAll { $0 == (decimalIsDot ? "," : ".") }
            if !decimalIsDot { s = s.replacingOccurrences(of: ",", with: ".") }
        } else if lastComma != nil {
            // Only commas: a single comma with 1–2 trailing digits is a decimal
            // ("5,00"); anything else is grouping ("1,299" / "1,000,000").
            let parts = s.split(separator: ",", omittingEmptySubsequences: false)
            if parts.count == 2, parts[1].count <= 2 {
                s = s.replacingOccurrences(of: ",", with: ".")
            } else {
                s.removeAll { $0 == "," }
            }
        }

        guard let value = Double(s) else { return nil }
        return negative ? -abs(value) : value
    }

    /// Normalizes a free-text status to a valid `item_status`, else nil.
    static func status(_ raw: String?) -> String? {
        guard let raw else { return nil }
        let key = raw.lowercased().trimmingCharacters(in: .whitespaces)
        let table: [String: String] = [
            "sourced": "sourced", "acquired": "acquired", "bought": "acquired",
            "cataloged": "cataloged", "catalogued": "cataloged",
            "draft": "drafted", "drafted": "drafted",
            "listed": "listed", "active": "listed", "live": "listed",
            "sold": "sold",
            "shipped": "shipped",
            "returned": "returned",
            "keeping": "keeping", "wearing": "wearing",
        ]
        return table[key]
    }

    /// Normalizes a free-text category to a valid `item_category`, else nil.
    static func category(_ raw: String?) -> String? {
        guard let raw else { return nil }
        let key = raw.lowercased().filter { $0.isLetter }
        let table: [String: FlipdeskCategory] = [
            "clothing": .clothing, "clothes": .clothing, "apparel": .clothing,
            "shoes": .shoes, "sneakers": .shoes, "footwear": .shoes,
            "watches": .watches, "watch": .watches,
            "sportscards": .sportsCards, "cards": .sportsCards,
            "collectibles": .collectibles, "collectible": .collectibles,
            "electronics": .electronics,
            "books": .books, "book": .books,
            "jewelry": .jewelry, "jewellery": .jewelry, "ring": .jewelry,
            "necklace": .jewelry, "earrings": .jewelry, "bracelet": .jewelry,
            "bags": .bags, "bag": .bags, "handbag": .bags, "purse": .bags,
            "backpack": .bags, "tote": .bags,
            "accessories": .accessories, "accessory": .accessories,
            "hat": .accessories, "cap": .accessories, "belt": .accessories,
            "scarf": .accessories, "sunglasses": .accessories,
            "other": .other,
        ]
        return table[key]?.rawValue
    }

    /// Which number comes first in a slashed date (US-3271).
    ///
    /// ⚠ `dateISO` USED TO ASSUME MONTH-FIRST, ALWAYS. A UK or European sheet's
    /// `03/09/2026` means 3 September and parsed as 9 March: the wrong date, in
    /// the right shape, with no error and nothing dropped. `droppedCells`
    /// cannot see it either, because the cell parsed fine. It lands in the
    /// acquired date, which drives aging and every tax-year boundary.
    enum SlashOrder: Equatable {
        case monthFirst
        case dayFirst
    }

    /// Reads the order off the WHOLE COLUMN rather than one cell, which is what
    /// makes it decidable rather than a coin flip. `13/04/2026` can only be
    /// day-first; `04/13/2026` can only be month-first. One unambiguous value
    /// settles the column, because a spreadsheet column is written by one
    /// person in one format.
    ///
    /// Only when every value is ambiguous (both numbers 12 or under, all the
    /// way down) does the device locale decide, and the import preview then
    /// says which reading it used. Guessing quietly is what this replaces.
    static func slashOrder(in samples: [String], locale: Locale = .current) -> SlashOrder {
        var sawDayFirst = false
        var sawMonthFirst = false
        for sample in samples {
            guard let (first, second) = slashParts(sample) else { continue }
            if first > 12 { sawDayFirst = true }
            if second > 12 { sawMonthFirst = true }
        }
        // A column carrying both is broken whichever way we read it. Month-first
        // matches the previous behaviour, so this cannot make an existing import
        // worse, and the ambiguous-column notice still names the reading.
        if sawDayFirst && !sawMonthFirst { return .dayFirst }
        if sawMonthFirst { return .monthFirst }
        return localeOrder(locale)
    }

    /// The two leading numbers of a `d/m/y`-shaped value, or nil when the value
    /// is not slashed at all (an ISO date settles nothing).
    static func slashParts(_ raw: String) -> (Int, Int)? {
        let parts = raw.trimmingCharacters(in: .whitespaces).split(separator: "/")
        guard parts.count == 3,
              let first = Int(parts[0]), let second = Int(parts[1]),
              first > 0, second > 0
        else { return nil }
        return (first, second)
    }

    /// What this device's own locale puts first. `dateFormat(fromTemplate:)`
    /// returns "M/d/y" in the US and "dd/MM/y" in the UK, so the first of `d`
    /// or `M` to appear is the answer.
    static func localeOrder(_ locale: Locale) -> SlashOrder {
        guard let template = DateFormatter.dateFormat(
            fromTemplate: "yMd", options: 0, locale: locale
        ) else { return .monthFirst }
        let day = template.firstIndex(of: "d")
        let month = template.firstIndex(of: "M")
        guard let day, let month else { return .monthFirst }
        return day < month ? .dayFirst : .monthFirst
    }

    /// Parses a date in common spreadsheet shapes to an ISO `YYYY-MM-DD` string.
    ///
    /// `order` decides only how a slashed date is read; ISO input ignores it.
    static func dateISO(_ raw: String?, order: SlashOrder = .monthFirst) -> String? {
        guard let raw = raw?.trimmingCharacters(in: .whitespaces), !raw.isEmpty else { return nil }
        let slashed = order == .dayFirst
            ? ["dd/MM/yyyy", "d/M/yyyy", "dd/MM/yy", "d/M/yy"]
            : ["MM/dd/yyyy", "M/d/yyyy", "MM/dd/yy", "M/d/yy"]
        let formats = ["yyyy-MM-dd"] + slashed + ["yyyy/MM/dd"]
        let parser = DateFormatter()
        parser.locale = Locale(identifier: "en_US_POSIX")
        parser.timeZone = TimeZone(identifier: "UTC")
        for fmt in formats {
            parser.dateFormat = fmt
            if let date = parser.date(from: raw) {
                let out = DateFormatter()
                out.locale = Locale(identifier: "en_US_POSIX")
                out.timeZone = TimeZone(identifier: "UTC")
                out.dateFormat = "yyyy-MM-dd"
                return out.string(from: date)
            }
        }
        return nil
    }
}

File: android/app/src/main/java/com/gradethread/app/importer/ImportField.kt
package com.gradethread.app.importer

import androidx.annotation.StringRes
import com.gradethread.app.R

import com.gradethread.app.capture.FlipdeskCategory
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * US-1389 (iOS `ImportField`, web `src/lib/import-mapping.ts`): the fields a
 * spreadsheet column can map to.
 */
/**
 * US-2976: [label] is a string RESOURCE. These are the column choices in the
 * CSV importer's mapping step, so a seller matching their own spreadsheet to
 * our fields was reading fourteen English words to do it.
 */
enum class ImportField(@StringRes val label: Int) {
    SKIP(R.string.import_field_skip),
    SKU(R.string.import_field_sku),
    TITLE(R.string.import_field_title),
    BRAND(R.string.import_field_brand),
    STYLE(R.string.import_field_style),
    SIZE(R.string.import_field_size),
    COLOR(R.string.import_field_color),
    MATERIAL(R.string.import_field_material),
    CONDITION_NOTES(R.string.import_field_condition_notes),
    CATEGORY(R.string.import_field_category),
    PURCHASE_PRICE(R.string.import_field_purchase_price),
    LIST_PRICE(R.string.import_field_list_price),
    PURCHASE_DATE(R.string.import_field_purchase_date),
    STATUS(R.string.import_field_status),
    ;

    /** Title is the only required field — a row without one cannot be inserted. */
    val isRequired: Boolean get() = this == TITLE

    companion object {
        /**
         * Best guess from a header, at any casing or spacing.
         *
         * An unknown header maps to SKIP rather than to something plausible:
         * silently importing a "Location" column into "Notes" is worse than
         * making the seller point at it themselves.
         */
        fun guess(header: String): ImportField {
            val key = header.lowercase(Locale.US).filter { it.isLetterOrDigit() }
            return TABLE[key] ?: SKIP
        }

        private val TABLE: Map<String, ImportField> = mapOf(
            "item" to SKU, "itemnumber" to SKU, "itemno" to SKU, "sku" to SKU,
            "itemtitle" to TITLE, "title" to TITLE, "name" to TITLE,
            "brand" to BRAND,
            "style" to STYLE, "model" to STYLE,
            "size" to SIZE,
            "color" to COLOR, "colour" to COLOR,
            "material" to MATERIAL, "fabric" to MATERIAL,
            "notes" to CONDITION_NOTES, "note" to CONDITION_NOTES,
            "description" to CONDITION_NOTES, "desc" to CONDITION_NOTES,
            "category" to CATEGORY, "cat" to CATEGORY,
            "purchaseprice" to PURCHASE_PRICE, "cost" to PURCHASE_PRICE,
            "paid" to PURCHASE_PRICE,
            "listprice" to LIST_PRICE, "price" to LIST_PRICE,
            "targetprice" to LIST_PRICE,
            "purchasedate" to PURCHASE_DATE, "purchased" to PURCHASE_DATE,
            "acquired" to PURCHASE_DATE,
            "status" to STATUS,
        )
    }
}

/**
 * US-1389 (iOS `ImportValue`): turning a spreadsheet cell into a stored value.
 *
 * Pure, and the locale handling is the reason. A CSV can come from anywhere,
 * and the naive "strip everything but digits and dots" reading turns
 * "1.299,00" into 1.29900 — a hundred-fold error in a cost basis that then
 * flows into every profit figure the seller looks at (iOS US-1162).
 */
object ImportValue {

    fun price(raw: String?): Double? {
        if (raw == null) return null
        val trimmed = raw.trim()
        // Accounting negatives wrap the WHOLE value in parens. A parenthetical
        // note in a price cell ("$20 (sale)") must not flip the sign.
        val negative = trimmed.startsWith("(") && trimmed.endsWith(")")

        var s = raw.filter { it.isDigit() || it == '.' || it == ',' || it == '-' }
        if (s.isEmpty()) return null

        val lastDot = s.lastIndexOf('.')
        val lastComma = s.lastIndexOf(',')
        if (lastDot >= 0 && lastComma >= 0) {
            // Both present: the RIGHT-MOST is the decimal separator.
            val decimalIsDot = lastDot > lastComma
            s = if (decimalIsDot) {
                s.filter { it != ',' }
            } else {
                s.filter { it != '.' }.replace(',', '.')
            }
        } else if (lastComma >= 0) {
            // Only commas: one comma with 1-2 trailing digits is a decimal
            // ("5,00"); anything else is grouping ("1,299" / "1,000,000").
            val parts = s.split(',')
            s = if (parts.size == 2 && parts[1].length <= 2) {
                s.replace(',', '.')
            } else {
                s.filter { it != ',' }
            }
        }

        val value = s.toDoubleOrNull() ?: return null
        return if (negative) -kotlin.math.abs(value) else value
    }

    /**
     * A free-text status, normalized — or null.
     *
     * Null rather than a guess: an unrecognized status becomes the default
     * ("cataloged"), which is honest, whereas mapping "pending" to "listed"
     * would tell the seller stock is live when it isn't.
     */
    fun status(raw: String?): String? {
        val key = raw?.trim()?.lowercase(Locale.US) ?: return null
        return STATUSES[key]
    }

    fun category(raw: String?): String? {
        val key = raw?.lowercase(Locale.US)?.filter { it.isLetter() } ?: return null
        return CATEGORIES[key]?.wire
    }

    /**
     * A spreadsheet date as `YYYY-MM-DD`.
     *
     * Ambiguous two-digit years and day/month order are the reason the formats
     * are an explicit ordered list: "03/04/2026" is March 4th under the US
     * layouts a US-locale export uses, and guessing per-device locale would
     * make the same file import differently on two phones.
     */
    fun dateIso(raw: String?): String? {
        val text = raw?.trim().orEmpty()
        if (text.isEmpty()) return null
        FORMATS.forEach { pattern ->
            runCatching {
                val parsed = LocalDate.parse(
                    text,
                    DateTimeFormatter.ofPattern(pattern, Locale.US),
                )
                return parsed.format(DateTimeFormatter.ISO_LOCAL_DATE)
            }
        }
        return null
    }

    private val FORMATS = listOf(
        "yyyy-MM-dd",
        "MM/dd/yyyy",
        "M/d/yyyy",
        "MM/dd/yy",
        "M/d/yy",
        "yyyy/MM/dd",
    )

    private val STATUSES = mapOf(
        "sourced" to "sourced", "acquired" to "acquired", "bought" to "acquired",
        "cataloged" to "cataloged", "catalogued" to "cataloged",
        "draft" to "drafted", "drafted" to "drafted",
        "listed" to "listed", "active" to "listed", "live" to "listed",
        "sold" to "sold",
        "shipped" to "shipped",
        "returned" to "returned",
        "keeping" to "keeping", "wearing" to "wearing",
    )

    private val CATEGORIES = mapOf(
        "clothing" to FlipdeskCategory.CLOTHING, "clothes" to FlipdeskCategory.CLOTHING,
        "apparel" to FlipdeskCategory.CLOTHING,
        "shoes" to FlipdeskCategory.SHOES, "sneakers" to FlipdeskCategory.SHOES,
        "footwear" to FlipdeskCategory.SHOES,
        "watches" to FlipdeskCategory.WATCHES, "watch" to FlipdeskCategory.WATCHES,
        "sportscards" to FlipdeskCategory.SPORTS_CARDS,
        "cards" to FlipdeskCategory.SPORTS_CARDS,
        "collectibles" to FlipdeskCategory.COLLECTIBLES,
        "collectible" to FlipdeskCategory.COLLECTIBLES,
        "electronics" to FlipdeskCategory.ELECTRONICS,
        "books" to FlipdeskCategory.BOOKS, "book" to FlipdeskCategory.BOOKS,
        "jewelry" to FlipdeskCategory.JEWELRY, "jewellery" to FlipdeskCategory.JEWELRY,
        "ring" to FlipdeskCategory.JEWELRY, "necklace" to FlipdeskCategory.JEWELRY,
        "earrings" to FlipdeskCategory.JEWELRY, "bracelet" to FlipdeskCategory.JEWELRY,
        "bags" to FlipdeskCategory.BAGS, "bag" to FlipdeskCategory.BAGS,
        "handbag" to FlipdeskCategory.BAGS, "purse" to FlipdeskCategory.BAGS,
        "backpack" to FlipdeskCategory.BAGS, "tote" to FlipdeskCategory.BAGS,
        "accessories" to FlipdeskCategory.ACCESSORIES,
        "accessory" to FlipdeskCategory.ACCESSORIES,
        "hat" to FlipdeskCategory.ACCESSORIES, "cap" to FlipdeskCategory.ACCESSORIES,
        "belt" to FlipdeskCategory.ACCESSORIES, "scarf" to FlipdeskCategory.ACCESSORIES,
        "sunglasses" to FlipdeskCategory.ACCESSORIES,
        "other" to FlipdeskCategory.OTHER,
    )
}
